pulp/apis/sas_api.py

```diff
@@ -0,0 +1,623 @@
+                                run;
+                                """.format(
+                        warmstart=warmstart_str,
+                        postfix=postfix,
+                        mpsfile=tmpMps,
+                        proc=proc,
+                        maxLen=min(nameLen, MAX_NAME_LENGTH),
+                        options=opt_str,
+                        decomp=decomp_str,
+                        decompmaster=decompmaster_str,
+                        decompmasterip=decompmasterip_str,
+                        decompsubprob=decompsubprob_str,
+                        rootnode=rootnode_str,
+                    ),
+                    results="TEXT",
+                )
+            else:
+                # Since 9.4M6+ optlp/optmilp can read mps files directly
+                # TODO: Check whether there are limits for length of variable and constraint names
+                res = sas.submit(
+                    """
+                                option notes nonumber nodate nosource pagesize=max;
+                                {warmstart}
+                                proc {proc} mpsfile=\"{mpsfile}\" {options} primalout=primalout{postfix} dualout=dualout{postfix};
+                                {decomp}
+                                {decompmaster}
+                                {decompmasterip}
+                                {decompsubprob}
+                                {rootnode}
+                                run;
+                                """.format(
+                        warmstart=warmstart_str,
+                        postfix=postfix,
+                        proc=proc,
+                        mpsfile=tmpMps,
+                        options=opt_str,
+                        decomp=decomp_str,
+                        decompmaster=decompmaster_str,
+                        decompmasterip=decompmasterip_str,
+                        decompsubprob=decompsubprob_str,
+                        rootnode=rootnode_str,
+                    ),
+                    results="TEXT",
+                )
+
+            self._delete_tmp_files(tmpMps, tmpMst)
+
+            # Store log and ODS output
+            self._log = res["LOG"]
+            if self.msg:
+                print(self._log)
+            self._macro = dict(
+                (key.strip(), value.strip())
+                for key, value in (
+                    pair.split("=") for pair in sas.symget("_OR" + proc + "_").split()
+                )
+            )
+
+            primal_out = sas.sd2df(f"primalout{postfix}")
+            dual_out = sas.sd2df(f"dualout{postfix}")
+
+
+            if self._macro.get("STATUS", "ERROR") != "OK":
+                raise PulpSolverError("PuLP: Error ({err_name}) \
+                        while trying to solve the instance: {name}".format(
+                        err_name=self._macro.get("STATUS", "ERROR"), name=lp.name))
+            status = self._read_solution(lp, primal_out, dual_out)
```

Comment:
what if there are no values available? should we read the soluton still?

Reply:
If no values available, the solution will be NaN.